### PR TITLE
Support IP-in-subnet queries in lookup tables

### PR DESCRIPTION
**Pull request by mavam**
This PR enhances lookup tables when their key type is a `subnet` value. In this case, it is now possible to probe the lookup table with a value of type `ip` and perform a *prefix match* on all contained subnets.

### CODE REVIEW

**Comment by dominiklohmann** [comment]
@mavam Out of curiosity: Did you see https://github.com/zeek/pysubnettree before and consider using the patricia trie from it?

**Comment by mavam** [comment]
Yep, this is exactly the implementation I'm using now.

File: libtenzir/builtins/contexts/lookup_table.cpp [hunk 1]
```diff
@@ -276,11 +296,25 @@ class ctx final : public virtual context {
 
   /// Inspects the context.
   auto show() const -> record override {
-    return record{{"num_entries", context_entries.size()}};
+    // TODO: there's not size() function for the patricia trie, so we need to
+    // track the number of elements externally and integrate it into this
+    // output.
```
Comment (mavam) [inline libtenzir/builtins/contexts/lookup_table.cpp:301]: We could track insertions/deletions and then compute the number of elements in the subnet tree.

File: libtenzir/builtins/contexts/lookup_table.cpp [hunk 2]
```diff
@@ -379,9 +419,11 @@ class ctx final : public virtual context {
 
   auto reset() -> caf::expected<void> override {
     context_entries.clear();
+    subnet_entries.clear();
     return {};
   }
 
+  // TODO: persist patricia trie.
```
Comment (mavam) [inline libtenzir/builtins/contexts/lookup_table.cpp:426]: @tenzir/pipelines-squad does this requires bumping the Flatbuffer version?

File: libtenzir/builtins/contexts/lookup_table.cpp [hunk 3]
```diff
@@ -415,8 +457,10 @@ class ctx final : public virtual context {
 
 private:
   map_type context_entries;
+  detail::subnet_tree subnet_entries;
 };
 
+// TODO: do we need a v2 loader now with the subnet optimization?
```
Comment (mavam) [inline libtenzir/builtins/contexts/lookup_table.cpp:463]: @tenzir/pipelines-squad what about this? 👆

File: libtenzir/builtins/contexts/lookup_table.cpp (hunk 2 above)
Comment (dominiklohmann) [reply libtenzir/builtins/contexts/lookup_table.cpp:426]: Why would you need to persist it at all? All the data is already present in the lookup table, so when loading you can just recreate the trie.

File: libtenzir/builtins/contexts/lookup_table.cpp (hunk 3 above)
Comment (dominiklohmann) [reply libtenzir/builtins/contexts/lookup_table.cpp:463]: Same as above—if you avoid persisting the patricia trie in the first place this is irrelevant.

File: libtenzir/builtins/contexts/lookup_table.cpp [hunk 4]
```diff
@@ -324,8 +358,14 @@ class ctx final : public virtual context {
     while (key_it != key_values.end()) {
       TENZIR_ASSERT(context_it != context_values.end());
       auto materialized_key = materialize(*key_it);
-      context_entries.insert_or_assign(materialized_key,
-                                       materialize(*context_it));
+      // Subnets never make it into the regular map of entries.
```
Comment (dominiklohmann) [inline libtenzir/builtins/contexts/lookup_table.cpp:361]: I think this makes everything a lot more complicated than it needs to be. You can remove a lot of the code in this file if you just add it to both.

Comment (mavam) [reply libtenzir/builtins/contexts/lookup_table.cpp:361]: But that 2x'es the space utilization.

Comment (dominiklohmann) [reply libtenzir/builtins/contexts/lookup_table.cpp:361]: I think this is irrelevant until someone complains about it. Let's wait with improving this until someone wants to add a few million subnets.

Comment (mavam) [reply libtenzir/builtins/contexts/lookup_table.cpp:361]: Okay. YOLO but fine with me for the first version.

Comment (dominiklohmann) [reply libtenzir/builtins/contexts/lookup_table.cpp:361]: Even then, if you really want to optimize it, you could still keep it separate in memory but persist it together as one lookup table. No need to change the on-disk format either way.

**Pull request merged**